proposed/vector-instruction-set-default.md

```diff
@@ -0,0 +1,100 @@
+
+It is useful to look at hardware dates again. Continuing with the Intel focus, we can see the following delivery of SIMD instructions in chips:
+
+* 2006: SSE4 was introduced with Core 2 x64 chips at the initial [Intel Core architecture](https://en.wikipedia.org/wiki/Intel_Core_(microarchitecture)) launch.
+* 2008: SSE4.2 was introduced with [Nehalem](https://en.wikipedia.org/wiki/Nehalem_(microarchitecture)) chips.
+* 2011: AVX was introduced with [Sandy Bridge](https://en.wikipedia.org/wiki/Sandy_Bridge) chips.
+* 2013: AVX2 was introduced with [Haswell](https://en.wikipedia.org/wiki/Haswell_(microarchitecture)) chips.
+
+Let's take a look at each operating system, assuming the Intel ISA.
+ 
+* On Windows, we need to be conservative. There are a lot of .NET users on Windows, and the majority of .NET desktop apps target Windows. Windows 10 requires AVX, but Win7 requires only SSE2. It is reasonable to expect that we'll continue supporting Windows 7 with .NET 6.0 (and probably not after that).
+* On [macOS](https://en.wikipedia.org/wiki/MacOS_version_history#Version_10.13:_%22High_Sierra%22), it appears that [Mac machines have had AVX2 since late 2013](https://en.wikipedia.org/wiki/List_of_Macintosh_models_grouped_by_CPU_type#Haswell), when they adopted [Haswell chips](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2). In terms of macOS, [macOS Big Sur (11.0)](https://support.apple.com/kb/sp833?locale=en_US) appears to be the first version to require Haswell. .NET 5.0, for example, supports [macOS High Sierra (10.13)](https://support.apple.com/kb/SP765?locale=en_US) and later. macOS 10.13 supports hardware significantly before Haswell. For .NET 6.0, we'll likely continue to choose a "10." macOS version as the minimum, possibly [macOS Catalina (10.15)](https://support.apple.com/kb/SP803?locale=en_US), aligning with our past practice of slowly (not aggressively) moving forward OS requirements. Zooming out, macOS x64 is now a legacy platform. We shouldn't rock the boat with a change that could cause a performance regression at this late stage of the macOS x64 lifecycle.
+* On Linux, we can assume that there is at least as much diversity of hardware as Windows, however, we can also assume that .NET usage on Linux is more narrow, targeted at developers and (mostly) production deployments.
+
+The following is a draft plan on how to approach different systems for this proposal:
+
+* Windows x86 -- Target SSE2 (status-quo; match Windows 7)
+* Windows x64 -- Target AVX2
+* Linux x64 -- Target AVX2
+* macOS x64 -- Target SSE2 (status-quo; alternatively, target SSE4 if it is straightforward and has significant value)
```

Comment:
> target SSE4 if it is straightforward and has significant value

I'd drop this alternative part and just leave it with SSE2.

Reply:
I don't disagree, but could you explain why? It is unlikely that anyone would notice. SSE4 has been in place for about 15 years.

Reply:
I think it's not worth the effort at this point and there is non-zero risk/cost associated with this (e.g. riskier cross-releases backports).

Reply:
I don't see the cross-release backport risk (this is just a crossgen default), but I agree we should stick with SSE2 and that's the recommendation. I think that's what we'll do. I'll delete the SSE4 in a later update.

Reply:
I'm not sure I agree.

SSE4.1 is a 13-15 year old baseline and is almost a requirement to get decent perf out of many of the current vectorized code paths in the BCL. Otherwise, you are missing fairly "basic" operations like blend, dot product, and align right, which are used in things ranging from the System.Numerics.Vector types to the text encoding/decoding algorithms backing `string`.
* It's so old that the x86 emulation layers in Windows and Apple Silicon are able to emulate it and make it "available"
* We don't even provide just Sse2 fallbacks for some of our code paths that use Ssse3/Sse41 because it would make the code more complicated for likely little practical benefit. Instead, users just get the "software" (non-vectorized) fallback. For example: https://source.dot.net/#http2cat/StringUtilities.cs,703

For macOS, in particular, 10.13 (High Sierra) is the oldest still maintained version. Looking at the hardware requirements, it is a Late 2009 MacBook at the worst: https://support.apple.com/kb/SP765?viewlocale=en_US&locale=en_US

Looking at the technical specs: https://support.apple.com/kb/SP548?locale=en_US, this is running a Core 2 Duo which has SSE4.1

While there may be some users on even older hardware or versions (potentially back to snow leopard) they certainly aren't supported by Apple nor will they likely be supported by .NET 6 based on what versions of macOS we currently support.

I would imagine similar exists for practically all OEM computers running Windows 7 or later. Windows 7 was released in 2009 and while there are likely some users that have upgraded to it from XP or Vista based machines, I would speculate those are likely few and far between in practice.

Reply:
Let's measure and see what the difference is. I agree that the compat risk here is very low.